### PR TITLE
DBZ-2383 OpenShift tests for DB2 connector

### PR DESCRIPTION
**Pull request by jcechace**
DB2 connector tests with OpenShift.

### CODE REVIEW

**Comment by jcechace** [comment]
@gunnarmorling  I still wasn't able to resolve the db query issues. So probably some assistance from somebody experienced with this DB is required.  https://github.com/debezium/debezium/pull/1725/files#diff-987e3d58c2080a07620b08131fe673efR54 point to the problematic part.

File: debezium-testing/debezium-testing-openshift/src/main/java/io/debezium/testing/openshift/tools/databases/db2/DB2Controller.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    private static final Logger LOGGER = LoggerFactory.getLogger(DB2Controller.class);
+    private static final String READINESS_SQL_SELECT = "SELECT 1 FROM DB2INST1.CUSTOMERS;";
+
+    public DB2Controller(Deployment deployment, List<Service> services, String dbType, OpenShiftClient ocp) {
+        super(deployment, services, dbType, ocp);
+    }
+
+
+    public void initialize() {
+        LOGGER.info("Waiting until DB2 instance is ready");
```
Comment (gunnarmorling) [inline debezium-testing/debezium-testing-openshift/src/main/java/io/debezium/testing/openshift/tools/databases/db2/DB2Controller.java:55]: Hey @SeanRooooney, we could need some helps on Db2 :) Specifically,@jcechace here is looking for a way to ensure the database has started (as a container) in integration tests before running any test. I.e. some readiness probe of sort. We do specific selects e.g. for Postgres in order to achieve that, and are looking for any Db2 equivalent. Thanks for any advise.

**Comment by jcechace** [comment]
Even the following snippet leads to the error bellow
```
try (Connection con = DriverManager.getConnection(getDatabaseUrl(), ConfigProperties.DATABASE_DB2_DBZ_USERNAME, ConfigProperties.DATABASE_DB2_DBZ_PASSWORD)) {
            try (Statement s = con.createStatement()) {
                s.execute("SELECT 1 FROM DB2INST1.CUSTOMERS");
            }
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
```

```
***** Out of Package Error Occurred (2020-07-28 17:48:03.419) *****

Exception stack trace:
com.ibm.db2.jcc.am.SqlException: DB2 SQL Error: SQLCODE=-805, SQLSTATE=51002, SQLERRMC=NULLID.SYSSH200 0X5359534C564C3031, DRIVER=4.27.25
    com.ibm.db2.jcc.am.b7.a(b7.java:815)
    com.ibm.db2.jcc.am.b7.a(b7.java:66)
    com.ibm.db2.jcc.am.b7.a(b7.java:140)
    com.ibm.db2.jcc.am.k6.c(k6.java:2825)
    com.ibm.db2.jcc.t4.ab.s(ab.java:1687)
    com.ibm.db2.jcc.t4.ab.i(ab.java:209)
    com.ibm.db2.jcc.t4.ab.b(ab.java:96)
    com.ibm.db2.jcc.t4.p.a(p.java:32)
    com.ibm.db2.jcc.t4.av.i(av.java:150)
    com.ibm.db2.jcc.am.k6.al(k6.java:2204)
    com.ibm.db2.jcc.am.k6.a(k6.java:3331)
    com.ibm.db2.jcc.am.k6.e(k6.java:1132)
    com.ibm.db2.jcc.am.k6.execute(k6.java:1111)
    io.debezium.testing.openshift.tools.databases.SqlDatabaseClient.lambda$execute$4(SqlDatabaseClient.java:71)
    io.debezium.testing.openshift.tools.databases.SqlDatabaseClient.doExecute(SqlDatabaseClient.java:40)
    io.debezium.testing.openshift.tools.databases.SqlDatabaseClient.lambda$execute$0(SqlDatabaseClient.java:50)
    org.awaitility.core.CallableCondition$ConditionEvaluationWrapper.eval(CallableCondition.java:99)
    org.awaitility.core.ConditionAwaiter$ConditionPoller.call(ConditionAwaiter.java:203)
    org.awaitility.core.ConditionAwaiter$ConditionPoller.call(ConditionAwaiter.java:190)
    java.base/java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:264)
    java.base/java.util.concurrent.FutureTask.run(FutureTask.java)
    java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1135)
    java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
    java.base/java.lang.Thread.run(Thread.java:844)

Concurrently open statements:
1. SQL string:  SELECT 1 FROM DB2INST1.CUSTOMERS;
   Number of statements:  1
********************
```

Note that this is the very first SQL statement and the error is thrown before any other is executed so none of the "Large amount of statements in loop" scenarios is applicable here.

File: debezium-testing/debezium-testing-openshift/src/main/java/io/debezium/testing/openshift/tools/databases/db2/DB2Controller.java (hunk 1 above)
Comment (jcechace) [reply debezium-testing/debezium-testing-openshift/src/main/java/io/debezium/testing/openshift/tools/databases/db2/DB2Controller.java:55]: The issue at hand is not exactly about detecting the DB startup. It's more about why the statement execution fails.

**Comment by SeanRooooney** [comment]
@zrlurb could you have a look ?

File: debezium-testing/debezium-testing-openshift/src/main/java/io/debezium/testing/openshift/tools/databases/db2/DB2Controller.java (hunk 1 above)
Comment (zrlurb) [reply debezium-testing/debezium-testing-openshift/src/main/java/io/debezium/testing/openshift/tools/databases/db2/DB2Controller.java:55]: it looks like not in a readiness state. You can connect, but not all of the preparing work is done. About IBM Db2 on Openshift will be probably found here https://hub.docker.com/r/ibmcom/db2 .

**Comment by zrlurb** [comment]
It looks like the Db2 is not ready started.

**Comment by gunnarmorling** [comment]
> It looks like the Db2 is not ready started.

Yes, that seems exactly to be the problem. Our question is: how can we establish some sort of barrier to await the start-up, so it's ready to receive queries, before executing the tests?

**Comment by lga-zurich** [comment]
It's not a great solution, but given that this is OpenShift, it has always served me well to wait for "Setup has completed" on the container/pod log to appear to know that the Db2 instance is up and running.

**Comment by jcechace** [comment]
Not sure if that will work for us since we are performing additional CDC setup -- which causes the db to stop and start again I will check the logs

**Comment by zrlurb** [comment]
What are the additional CDC setup?

**Comment by jcechace** [comment]
@zrlurb  https://github.com/debezium/debezium-examples/tree/master/tutorial/debezium-db2-init/db2server

**Comment by zrlurb** [comment]
@jcechace 
I have written the db2server Dockerfile. 
Docker build is inherited from ibmcom/db2 (Db2 Server).

You will find ADD sql and sh files to directories only in the extension of  ibmcom/db2 Dockerfile. In the end, you will find ADD cdcsetup.sh /var/custom. This script will be executed by the first docker startup of the db2 after the complete startup of the database.

All executable in the /var/custom will be called after db2 ist well started (read about ibmcom/db2 image).

**Comment by jcechace** [comment]
@zrlurb  Yes and the [cdcsetup.sh on line 16 ](https://github.com/debezium/debezium-examples/blob/master/tutorial/debezium-db2-init/db2server/cdcsetup.sh#L16) creates /asncdctools/src/asncdc.nlk

Which is exactly what is checked by [my readiness probe](https://github.com/debezium/debezium/pull/1725/files#diff-65b193ad7e65c1745464ae3b94171ee5R51)

Still have the issue described above

**Comment by zrlurb** [comment]
@jcechace 
you could try this:

add the following lines at the end of db2setup.sh :
```
rc=1
echo "Waiting for DB2 start ( $DBNAME ) ."
while [ "$rc" -ne 0 ]
do
   sleep 5
   db2 connect to $DBNAME
   rc=$?
   echo '.'
done
```

**Comment by jcechace** [comment]
@zrlurb  I can confirm the paths. I really doubt the issue at hand (inability to execute query via JDBC) is caused by the database not being ready (since it allows me to execute the same query via cmd client from within the pod). 

Nevertheless, I will recheck this once I have resolved a new issue with strimzi.

**Comment by jcechace** [comment]
So I can confirm that the execution doesn't work even though the db is ready (at least for local cmd client). Is there perhaps some configuration which causes remotely executed queries to fail? Although I'd find the error message cryptic as hell in that case.

**Comment by jcechace** [comment]
So I managed to confirm that it is possible to connect to the running db2 instance and execute a query remotely from db2 cmd client running at another node. Still, no idea why it is not possible from jdbc connection though.

**Comment by jcechace** [comment]
The entire reproducer

```
package io.debezium.testing.openshift.db2;

import io.debezium.testing.openshift.tools.ConfigProperties;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;
import java.sql.Statement;

/**
 * @author Jakub Cechacek
 */
public class DB2ConnectionIT  {
    
    public static final String URL = "jdbc:db2://a75e818538a5f41c3aefaefc51cd68f4-181580172.eu-west-1.elb.amazonaws.com:50000/TESTDB";
    public static final String USERNAME = "db2inst1";
    public static final String PASSWORD = "=Password!";

    @BeforeAll
    public static void setupDatabase() throws IOException, InterruptedException, ClassNotFoundException {
        Class.forName("com.ibm.db2.jcc.DB2Driver");
    }

    @Test
    public void shouldExecuteQuery() {
        try (Connection con = DriverManager.getConnection(URL, USERNAME, PASSWORD)) {
            try (Statement s = con.createStatement()) {
                s.execute("SELECT 1 FROM DB2INST1.CUSTOMERS");
            }
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }
}
```

```
***** Out of Package Error Occurred (2020-08-07 00:16:44.893) *****

Exception stack trace:
com.ibm.db2.jcc.am.SqlException: DB2 SQL Error: SQLCODE=-805, SQLSTATE=51002, SQLERRMC=NULLID.SYSSH200 0X5359534C564C3031, DRIVER=4.27.25
    com.ibm.db2.jcc.am.b7.a(b7.java:815)
    com.ibm.db2.jcc.am.b7.a(b7.java:66)
    com.ibm.db2.jcc.am.b7.a(b7.java:140)
    com.ibm.db2.jcc.am.k6.c(k6.java:2825)
    com.ibm.db2.jcc.t4.ab.s(ab.java:1687)
    com.ibm.db2.jcc.t4.ab.i(ab.java:209)
    com.ibm.db2.jcc.t4.ab.b(ab.java:96)
    com.ibm.db2.jcc.t4.p.a(p.java:32)
    com.ibm.db2.jcc.t4.av.i(av.java:150)
    com.ibm.db2.jcc.am.k6.al(k6.java:2204)
    com.ibm.db2.jcc.am.k6.a(k6.java:3331)
    com.ibm.db2.jcc.am.k6.e(k6.java:1132)
    com.ibm.db2.jcc.am.k6.execute(k6.java:1111)
    io.debezium.testing.openshift.db2.DB2ConnectionIT.shouldExecuteQuery(DB2ConnectionIT.java:36)
    java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    java.base/java.lang.reflect.Method.invoke(Method.java:564)
    org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:675)
    org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
    org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:125)
    org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:132)
    org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:124)
    org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:74)
    org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
    org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
    org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:104)
    org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:62)
    org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:43)
    org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:35)
    org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
    org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
    org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$6(TestMethodTestDescriptor.java:202)
    org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
    org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:198)
    org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:135)
    org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:69)
    org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:135)
    org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
    org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
    org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
    org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
    org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
    org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
    org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
    java.base/java.util.ArrayList.forEach(ArrayList.java:1378)
    org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
    org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
    org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
    org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
    org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
    org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
    org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
    org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
    org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
    java.base/java.util.ArrayList.forEach(ArrayList.java:1378)
    org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
    org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
    org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
    org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
    org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
    org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
    org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
    org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
    org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
    org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:32)
    org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
    org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:51)
    org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:229)
    org.junit.platform.launcher.core.DefaultLauncher.lambda$execute$6(DefaultLauncher.java:197)
    org.junit.platform.launcher.core.DefaultLauncher.withInterceptedStreams(DefaultLauncher.java:211)
    org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:191)
    org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:128)
    com.intellij.junit5.JUnit5IdeaTestRunner.startRunnerWithArgs(JUnit5IdeaTestRunner.java:74)
    com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
    com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
    com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)

Concurrently open statements:
1. SQL string:  SELECT 1 FROM DB2INST1.CUSTOMERS
   Number of statements:  1
********************
```

**Comment by jcechace** [comment]
Using the following jdbc driver

```xml
<dependency>
    <groupId>com.ibm.db2</groupId>
    <artifactId>jcc</artifactId>
    <version>11.5.4.0</version>
</dependency>
```

**Comment by zrlurb** [comment]
@jcechace , read, please. Hopefully, it will help you, to find a solution
https://www.ibm.com/support/pages/node/385983
http://db2commerce.com/betasite3/2011/04/15/sql0805/

**Comment by jcechace** [comment]
@zrlurb 

```
The preferred way to address this issue is to modify the application code so that it properly closes statement objects when finished using them.
```
Please have a look at the code I posted. I am pretty sure that it closes everything imaginable and no statements are being leaked. Not to mention it is just a single statement.

**Comment by jcechace** [comment]
Ok the same code works with no issue when run against DB run locally from [tutorial's docker-compose](https://github.com/debezium/debezium-examples/blob/master/tutorial/docker-compose-db2.yaml). It fails with the  quay.io/debezium/db2-cdc:1.2 deployed in OCP

**Comment by zrlurb** [comment]
@jcechace 
is it possible to try following:

could you add the line in the script debezium-examples/tutorial/debezium-db2-init/db2server/dbsetup.sh

add line 50: 
    db2 terminate

so it is absolutely clear "no hanging statement" from preconfiguration

**Comment by jcechace** [comment]
Unfortunately no difference 

```
Concurrently open statements:
1. SQL string:  SELECT 1 FROM DB2INST1.CUSTOMERS
   Number of statements:  1
********************
```

However, I can now confidently say that the issue manifests only when the DB is running within openshift

**Comment by gunnarmorling** [comment]
As discussed during the call: @jcechace and @jpechane will do a shared session to progess with this.

**Comment by lga-zurich** [comment]
> As discussed during the call: @jcechace and @jpechane will do a shared session to progess with this.

FWIW, I did a simple deployment using the quay.io/debezium/db2-cdc:1.2 image on an OpenShift cluster on IBM Cloud, and I could connect just fine to the service and query the tables. I did that _once the setup scripts were done running_, of course. That took roughly 10 minutes.

If we can be of any help please ping.

**Comment by jcechace** [comment]
I just noticed these in the  pod log

```
2020-09-08-12.14.21.723852+000 E365548E1220          LEVEL: Event
PID     : 16219                TID : 140306821408512 PROC : db2sysc 0
INSTANCE: db2inst1             NODE : 000            DB   : TESTDB
APPHDL  : 0-52                 APPID: *LOCAL.DB2.200908121416
AUTHID  : DB2INST1             HOSTNAME: db2-7c89cfbf59-59x6v
EDUID   : 138                  EDUNAME: db2pcsd (TESTDB) 0
FUNCTION: DB2 UDB, access plan manager, sqlraCacheSpaceDaemon, probe:100
MESSAGE : ADM4500W  A package cache overflow condition has occurred. There is 
          no error but this indicates that the package cache has exceeded the 
          configured maximum size. If this condition persists, you should 
          perform additional monitoring to determine if you need to change the 
          PCKCACHESZ DB configuration parameter. You could also set it to 
          AUTOMATIC.
REPORT  : APM : Package Cache : info
IMPACT  : Unlikely
DATA #1 : String, 268 bytes
Package Cache Overflow
memory needed             : 959295
current used size (OSS)   : 2153458
maximum cache size (APM)  : 1271398
maximum logical size (OSS): 2259180
maximum used size (OSS)   : 2621440
owned size (OSS)          : 2621440
number of overflows       : 4
```

I suspect there might be some configuration difference on the side of OCP. @lga-zurich  could you share your deployment descriptor?

**Comment by Naros** [comment]
@lga-zurich, if a query returns error -805 and message `NULLID.SYSSH200`, how should we cope with that?  Can you explain what that means in DB2 terms so we can understand how we're suppose to react to such an error?  For example, does this mean something isn't yet ready or the operation should be retried?  Does this mean some poor configuration that should be adjusted?

**Comment by jcechace** [comment]
Previously mentioned error appear also when the value is actually set to automatic

```
[db2inst1@db2-7c89cfbf59-xcqmn /]$ db2 get db cfg for TESTDB | grep "PCKCACHESZ"
 Package cache size (4KB)                   (PCKCACHESZ) = AUTOMATIC(16384)
```

**Comment by lga-zurich** [comment]
I am not performing any configuration on the Db2 instance. This is the deployment I'm using, note that I set this container as _privileged_.
```
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: db2
  labels:
    app: db2
spec:
  replicas: 1
  selector:
    matchLabels:
      app: db2
  strategy:
    rollingUpdate:
      maxSurge: 1
      maxUnavailable: 1
    type: RollingUpdate
  template:
    metadata:
      name: db2
      labels:
        app: db2
    spec:
      affinity:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
            - matchExpressions:
              - key: beta.kubernetes.io/arch
                operator: In
                values:
                - amd64
      containers:
      - env:
        - name: LICENSE
          value: accept
        - name: DB2INST1_PASSWORD
          value: XXX
        - name: DBNAME
          value: DEMODB
        image: quay.io/debezium/db2-cdc:1.2
        imagePullPolicy: IfNotPresent
        name: db2
        resources:
          requests:
            cpu: 500m
            memory: 8Gi
        securityContext:
          privileged: true
        volumeMounts:
        - mountPath: /database
          name: db2-db
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      terminationGracePeriodSeconds: 30
      volumes:
      - name: db2-db
        persistentVolumeClaim:
          claimName: db2-db-pvc
```

**Comment by gunnarmorling** [comment]
@jcechace, any news on this as per your latest insights?

**Comment by jcechace** [comment]
We should be good to merge this. For whatever reason, the privileged scc needs to be setup with desired project active (unlike say anyuuid) even though the SA where it is set is actually fully classified with project name...  unless this happened the scc was sometimes not applied (behaviour differed based on OCP version and probably also weather).  Anyway the nightmare seems to be over.

**Comment by jcechace** [comment]
Ok it's not.. the error above still sometimes happens, there will be something else...  So  I will try to look into few more things.  we are currently at the state where it looks like the first deployment of the database is successful but it degrades over time.

**Comment by jcechace** [comment]
Found the remaining issue -- the wait time when recreating database pods was too low. At the same time, it looks like the out of package issue described above is indeed caused by a connection before the database is fully ready -- however it puts the database into state from which it never recoveres.  

I still don't believe this solution is 100% error prone but it seems to work quite reliably as I managed to have 6 successful test runs in a row.

**Comment by gunnarmorling** [comment]
Ok, thanks a lot, will merge it then. The latest test failure is unrelated to this change. I'll resolve the issue, if there's problems remaining, let's take it from there. Thanks, @jcechace and @SeanRooooney et. all for your help!

**Pull request merged**